src/executables/elm_ats_api/elm_ats_driver.cc

```diff
@@ -0,0 +1,447 @@
+      subsurf_ss[0][col_iter[i]] = root_transpiration[col*ncol_cells_+i] * sub_mol_h20_kg;
+    }
+  }
+  // mark sources as changed
+  changedEvaluatorPrimary(srf_src_key_, Amanzi::Tags::NEXT, *S_);
+  changedEvaluatorPrimary(sub_src_key_, Amanzi::Tags::NEXT, *S_);
+}
+
+
+void
+ELM_ATSDriver::get_waterstate(double *surface_pressure, double *soil_pressure, double *saturation, int *ncols, int *ncells)
+{
+
+  S_->GetEvaluator(pd_key_, Amanzi::Tags::NEXT).Update(*S_, pd_key_);
+  const Epetra_MultiVector& pd = *S_->Get<Amanzi::CompositeVector>(pd_key_, Amanzi::Tags::NEXT)
+      .ViewComponent("cell", false);
+  S_->GetEvaluator(pres_key_, Amanzi::Tags::NEXT).Update(*S_, pres_key_);
+  const Epetra_MultiVector& pres = *S_->Get<Amanzi::CompositeVector>(pres_key_, Amanzi::Tags::NEXT)
+      .ViewComponent("cell", false);
+  S_->GetEvaluator(satl_key_, Amanzi::Tags::NEXT).Update(*S_, satl_key_);
+  const Epetra_MultiVector& sat = *S_->Get<Amanzi::CompositeVector>(satl_key_, Amanzi::Tags::NEXT)
+      .ViewComponent("cell", false);
+
+  AMANZI_ASSERT(*ncols == ncolumns_ == pd.MyLength());
+  AMANZI_ASSERT(*ncells == ncolumns_ * ncol_cells_);
+  AMANZI_ASSERT(*ncells == pres.MyLength());
+
+  for (Amanzi::AmanziMesh::Entity_ID col=0; col!=ncolumns_; ++col) {
+    surface_pressure[col] = pd[0][col];
+
+    auto& col_iter = mesh_subsurf_->cells_of_column(col);
+    for (std::size_t i=0; i!=col_iter.size(); ++i) {
+      soil_pressure[col*ncol_cells_+i] = pres[0][col_iter[i]];
+      saturation[col*ncol_cells_+i] = sat[0][col_iter[i]];
+    }
+  }
+
+}
+
+void ELM_ATSDriver::get_mesh_info(int *ncols_local, int *ncols_global, int *ncells_per_col,
+    double *dz, double *depth, double *elev, double *surf_area_m2, double *lat, double *lon)
+{
+  *ncols_local = static_cast<int>(mesh_surf_->num_entities(Amanzi::AmanziMesh::Entity_kind::CELL, Amanzi::AmanziMesh::Parallel_type::OWNED));
+  *ncols_global = static_cast<int>(mesh_surf_->num_entities(Amanzi::AmanziMesh::Entity_kind::CELL, Amanzi::AmanziMesh::Parallel_type::ALL));
+  *ncells_per_col = static_cast<int>(ncol_cells_);
+
+  col_depth(dz, depth);
+  changedEvaluatorPrimary("dz", Amanzi::Tags::NEXT, *S_);
+  changedEvaluatorPrimary("depth", Amanzi::Tags::NEXT, *S_);
+
+  S_->GetEvaluator(elev_key_, Amanzi::Tags::NEXT).Update(*S_, elev_key_);
+  const Epetra_MultiVector& elev_ats = *S_->Get<Amanzi::CompositeVector>(elev_key_, Amanzi::Tags::NEXT)
+      .ViewComponent("face",false);
+
+  for (Amanzi::AmanziMesh::Entity_ID col=0; col!=ncolumns_; ++col) {
+    // -- get the surface cell's equivalent subsurface face
+    Amanzi::AmanziMesh::Entity_ID f = mesh_surf_->entity_get_parent(Amanzi::AmanziMesh::CELL, col);
+    surf_area_m2[col] = mesh_subsurf_->face_area(f);
+    elev[col] = elev_ats[0][col];
+  }
+
+  // dummy lat lon for now
+  *lat = 0.5;
+  *lon = 0.5;
+}
+
+
+// helper function for collecting column dz and depth
+void ELM_ATSDriver::col_depth(double *dz, double *depth) {
+
```

Comment:
Is this going to be used by ELM or by ATS?  Assuming it is for ELM, isn't it easier to call the ATS mesh functions to get this info?

If it is for ATS, eventually we should have an "extruding" mesh constructor that would take dz.  But that's a distant eventually...

Reply:
This function and the calling function get_mesh_info() are very much a WIP. I was never 100% clear on the requirements for these functions and don't think they weren't used for the earlier coupled runs.